### ISSUE TITLE
[doc] Clearer getting-started documentation

### ISSUE THREAD

**Issue by jadephilipoom** [lowRISC/opentitan#11217]
Related to #5181

I've just been helping some collaborators get set up building OpenTitan, and I think that some changes to the installation guide and getting started guides could really improve that setup experience, particularly for those who don't want to edit the hardware at all (like the collaborators I was working with). Some issues I noticed:
- The "Install Build Requirements" page (`docs/ug/install_instructions/index.md`) includes a long list of requirements _everyone_ needs, which is then followed by detailed instructions for installing OpenOCD, Verilator, Verible, and Vivado. If you just proceed down the page, it appears that you need all of them, while new users probably need at most one of Verilator and Vivado and those who are not touching the RTL don't appear to need OpenOCD and Verible -- since those are both installed from source, it would be good to skip them if possible.
- For a new user, it's not clear where to go once the installation instructions page is complete. The answer for our case was "Getting Started Building Software" user guide followed by the "Getting Started with Verilator" guide, but it isn't at all obvious if you don't understand the project structure.
- If `/tools/riscv` doesn't work as an installation location for the toolchain, there's not really an explanation of how to install in a different location (although the docs do mention offhand that an `--install-dir` option exists, there's no clear "you should use it, here's an example" or explanation that you should also then set the `TOOLCHAIN_PATH` environment variable).
- The installation instructions don't seem to work out of the box on MacOS -- in particular, the `/tools/riscv` installation downloads compiled executables that can't be run due to `exec format error`s.

I propose the following changes:
- Make the install guide mostly a sequence of links to other pages and extend it, so it's more consise and people can follow it top-to-bottom from zero to running the hello-world program. See below for details.
- Clear up instructions about how to install from a non-default toolchain location, and standardize everything (including `get_toolchain.py`) to use the `TOOLCHAIN_PATH` environment variable so that it's consistent. Make sure to include "add this variable to your `.bashrc`".
- Figure out if the toolchain executables can work somehow on MacOS -- if so, add instructions, and if not, put a clear "MacOS not supported" message both there and at the very start of the install guide. Probably need to do the same check for Windows.

For the overarching install guide page, I'm imagining a structure like:
- Hi and welcome, explain that the installation will vary depending on how exactly you will interact with and/or change OpenTitan (in particular, whether you're editing the RTL, and whether you have an FPGA).
- Start: install the universal dependencies from system package manager and pip
- Install the riscv toolchain
  - If you already have one, link to the "bring your own toolchain" from the "Getting Started with Software" guide (which imho should be a separate page for everything custom-toolchain-related)
  - Otherwise, show how to use `get_toolchain.py`
- Explain when OpenOCD is needed (?), link to installation instructions
- Explain when Verible is needed (if you need to lint RTL), link to installation instructions
- Explain how to choose whether you want Verilator or Vivado, link to instructions for each
- If you only want to do design verification, you can stop here and go to the dedicated guide for that (?)
- Link to the build-software guide
- Link to the getting-started guides for verilator and FPGA, and guide the user to pick which one applies to them. Each should progress smoothly from this point to "hello world".

A couple things I'm still not sure about:
- When exactly do people need OpenOCD? I apparently never installed it (I think it didn't work when I first went through the install guide and I just decided to proceed and see if something broke, then nothing did), so at least for pure-software it doesn't seem necessary.
- What installation steps are applicable to design verification -- have I put the "go to dedicated guide" in the right place, or are steps beneath that necessary?
- Does the toolchain work at all on MacOS/Windows? If not, is it OK for the user to install whatever RISC-V toolchain does work for them and point `TOOLCHAIN_PATH` to there?

**Comment by timothytrippel** [comment]
> Related to #5181
> 
> I've just been helping some collaborators get set up building OpenTitan, and I think that some changes to the installation guide and getting started guides could really improve that setup experience, particularly for those who don't want to edit the hardware at all (like the collaborators I was working with). Some issues I noticed:
> 
> * The "Install Build Requirements" page (`docs/ug/install_instructions/index.md`) includes a long list of requirements _everyone_ needs, which is then followed by detailed instructions for installing OpenOCD, Verilator, Verible, and Vivado. If you just proceed down the page, it appears that you need all of them, while new users probably need at most one of Verilator and Vivado and those who are not touching the RTL don't appear to need OpenOCD and Verible -- since those are both installed from source, it would be good to skip them if possible.
> * For a new user, it's not clear where to go once the installation instructions page is complete. The answer for our case was "Getting Started Building Software" user guide followed by the "Getting Started with Verilator" guide, but it isn't at all obvious if you don't understand the project structure.
> * If `/tools/riscv` doesn't work as an installation location for the toolchain, there's not really an explanation of how to install in a different location (although the docs do mention offhand that an `--install-dir` option exists, there's no clear "you should use it, here's an example" or explanation that you should also then set the `TOOLCHAIN_PATH` environment variable).
> * The installation instructions don't seem to work out of the box on MacOS -- in particular, the `/tools/riscv` installation downloads compiled executables that can't be run due to `exec format error`s.
> 
> I propose the following changes:
> 
> * Make the install guide mostly a sequence of links to other pages and extend it, so it's more consise and people can follow it top-to-bottom from zero to running the hello-world program. See below for details.
> * Clear up instructions about how to install from a non-default toolchain location, and standardize everything (including `get_toolchain.py`) to use the `TOOLCHAIN_PATH` environment variable so that it's consistent. Make sure to include "add this variable to your `.bashrc`".
> * Figure out if the toolchain executables can work somehow on MacOS -- if so, add instructions, and if not, put a clear "MacOS not supported" message both there and at the very start of the install guide. Probably need to do the same check for Windows.
> 
> For the overarching install guide page, I'm imagining a structure like:
> 
> * Hi and welcome, explain that the installation will vary depending on how exactly you will interact with and/or change OpenTitan (in particular, whether you're editing the RTL, and whether you have an FPGA).
> * Start: install the universal dependencies from system package manager and pip
> * Install the riscv toolchain
>   
>   * If you already have one, link to the "bring your own toolchain" from the "Getting Started with Software" guide (which imho should be a separate page for everything custom-toolchain-related)
>   * Otherwise, show how to use `get_toolchain.py`
> * Explain when OpenOCD is needed (?), link to installation instructions
> * Explain when Verible is needed (if you need to lint RTL), link to installation instructions
> * Explain how to choose whether you want Verilator or Vivado, link to instructions for each
> * If you only want to do design verification, you can stop here and go to the dedicated guide for that (?)
> * Link to the build-software guide
> * Link to the getting-started guides for verilator and FPGA, and guide the user to pick which one applies to them. Each should progress smoothly from this point to "hello world".
> 
> A couple things I'm still not sure about:
> 
> * When exactly do people need OpenOCD? I apparently never installed it (I think it didn't work when I first went through the install guide and I just decided to proceed and see if something broke, then nothing did), so at least for pure-software it doesn't seem necessary.

The main purpose is to be able to use a debugger (like GDB) to step through software on-chip. For example, I have used it to step through a program running on top of a verilator sim (which is nice if you want to narrow down the specific point in a waveform you should be looking at to debug; e.g., I used it in #9508). However, it's real utility will be when we are debugging during bring-up (or even on the fpga) where we don't have access to waveforms. Last I used it though it was flaky at best (I think it was due to some JTAG issues that have been fixed since then).

> * What installation steps are applicable to design verification -- have I put the "go to dedicated guide" in the right place, or are steps beneath that necessary?

Admittedly, this area needs some work, both in terms of adding more content and organizing existing content. The short answer is they are somewhat scattered depending on the type of DV, i.e. block vs. chip-level. Supporting libraries for doing both are [here](https://docs.opentitan.org/hw/dv/), each IP has an individual DV page (e.g., the [adc_ctrl](https://docs.opentitan.org/hw/ip/adc_ctrl/doc/dv/index.html)) for block level, and the chip-level Earlgrey DV document is [here](https://docs.opentitan.org/hw/top_earlgrey/doc/dv/). Additionally, SW details on writing chip-level tests (including OTTF information) is [here](https://docs.opentitan.org/sw/device/lib/testing/). However I have a more detailed "Getting Started with SW Chip-Level Testing" document that I am trying to migrate to the website sometime soon, that should consolidate the chip-level DV information / fill in any gaps. In summary, this area needs work.

> * Does the toolchain work at all on MacOS/Windows? If not, is it OK for the user to install whatever RISC-V toolchain does work for them and point `TOOLCHAIN_PATH` to there?

I believe the short answer here is only that Ubuntu 18.04 is officially supported, though some other linux distros also work (AFAIK). If you are asking if **_any RISC-V toolchain_** will suffice (i.e. not necessarily the lowRISC built one, but one built for say windows), I do not think so because of the patches (i.e. the batmanip patch) applied to the lowRISC one, but perhaps @luismarques could comment on that? Additionally, from [here](https://github.com/lowRISC/lowrisc-toolchains), it also looks like the pre-built lowRISC toolchains are built with specific options.

-------

Additionally, a while back I proposed (#7260)  using the Docker container we provide as a quickstart mechanism to newcomers. I believe that lost traction on the fact that it wasn't officially tested in CI. If we could use the same container for CI then that would no longer be the case. Just a thought (more likely a stretch goal well beyond the scope of this proposal).

Overall, great proposal! I 100% agree we could improve this area of the project and think we would benefit from more clarity here, specifically in terms of scaling our efforts.
